### PR TITLE
Add Model registry api

### PR DESCRIPTION
**Pull request by manaswinidas**
Closes: https://issues.redhat.com/browse/RHOAIENG-2113

Collaborated with @dpanshug to create the model registry API and related tests

## Description

Adds model registry API ([Swagger link](https://editor.swagger.io/?url=https://raw.githubusercontent.com/opendatahub-io/model-registry/main/api/openapi/model-registry.yaml))

As for the APIs not included in this PR - their usage has not been determined yet(not in UX - [confirmed by the backend team](https://redhat-internal.slack.com/archives/C05QB0A582E/p1710322379688809?thread_ts=1710314042.265519&cid=C05QB0A582E)).

## How Has This Been Tested?

You can `console.log` the APIs to test 

## Test Impact

Added unit tests for the `custom.ts` and `errorUtils.ts`

## Request review criteria:

Self checklist (all need to be checked):
- [X] The developer has manually tested the changes and verified that the changes work
- [X] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [X] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [X] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 

- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

**Comment by manaswinidas** [comment]
WIP to add tests and testing failures

File: frontend/src/api/modelRegistry/errorUtils.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+const isError = (e: unknown): e is ModelRegistryError =>
+  ['code', 'message'].every((key) => key in (e as ModelRegistryError));
+
+export const handleModelRegistryFailures = <T>(promise: Promise<T>): Promise<T> =>
+  promise.then((result) => {
+    if (isError(result)) {
+      throw result;
+    }
+    return result;
+  });
```
Comment (andrewballantyne) [inline frontend/src/api/modelRegistry/errorUtils.ts:12]: Is this a copy of the Pipeline code or does Model Registry really send back errors as 200?

Comment (manaswinidas) [reply frontend/src/api/modelRegistry/errorUtils.ts:12]: Model Registry doesn't send errors as 200. This is what I'm testing currently

File: frontend/src/concepts/modelRegistry/types.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  UNKNOWN = 'UNKNOWN',
+  PENDING = 'PENDING',
+  LIVE = 'LIVE',
+  MARKED_FOR_DELETION = 'MARKED_FOR_DELETION',
+  DELETED = 'DELETED',
+  ABANDONED = 'ABANDONED',
+  REFERENCE = 'REFERENCE',
+}
+
+// TODO: add customProperties
```
Comment (manaswinidas) [inline frontend/src/concepts/modelRegistry/types.ts:36]: @lucferbux  can you help here to determine the type of the customProperties? [Swagger link](https://editor.swagger.io/?url=https://raw.githubusercontent.com/opendatahub-io/model-registry/main/api/openapi/model-registry.yaml)

File: frontend/src/api/modelRegistry/errorUtils.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      }
+      return result;
+    })
+    .catch((e) => {
+      if (isError(e)) {
+        throw new Error(e.message);
+      }
+      if (isCommonStateError(e)) {
+        // Common state errors are handled by useFetchState at storage level, let them deal with it
+        // TODO: check whether we need this or not
```
Comment (lucferbux) [inline frontend/src/api/modelRegistry/errorUtils.ts:21]: We'll circle back to this once we have a strong api endpoint

**Review by lucferbux — APPROVED** [review]
/lgtm

**Pull request merged**